### ISSUE TITLE
Idea: Symlink package contents instead of whole package

### ISSUE THREAD

**Issue by netanelgilad** [lerna/lerna#1321]
## Expected Behavior

When using lerna, I'd like the resulting workspace structure to resemble the way the packages would work with each other once installed using a package manager.

## Current Behavior

When bootstrapping a project with multiple packages, where package `a` is dependent on package `b`, then lerna creates a symlink from the node_modules of package `a` into the folder of package `b`. And the rest of the dependencies are installed properly for each of the packages. This creates an inconsistent behavior with the way package `a` would work when installed from a package manager, because in development package `b` would use it's own node_modules instead of the node_modules of package `a`. This causes problems when working with `peerDependencies` or problems where package `a` and `b` would use different instances of the same dependency even though they would have used the same instance with a proper install. 

## Possible Solution

A solution for that currently could be to use `--hoist` flag to dedup the dependencies of those packages, making sure they would use the same version.
But I would like to suggest a different solution, that I think would more closely reflect the way those packages would work when installed with a package manager. What if instead of symlinking the whole folder of package `b` (which means we also get the nested node_modules of package `b`) we would just symlink the files that would be published by the package (all files that are not `.npmignore`d). This would mean that we are no longer bringing the node_modules folder of package `b` into the scope of package `a`.
I haven't fully thought this idea through, and I'm creating this issue to start a discussion if this idea even makes sense 🤔. 
I'm sure that there are pitfalls that I haven't thought about, but this also requires a solution of how to install the dependencies of package `b` when installing the dependencies of package `a` because the latest `package.json` of package `b` isn't published, so this would require the package manager to know that in order to install package `b` it should look in the local `package.json` of package `b` and not in the registry (it may even not be published).

## Context

I'm currently working on project with a similar problem described in the example above. I can't use `--hoist` because of other reasons (https://github.com/yarnpkg/yarn/issues/4521, https://github.com/yarnpkg/yarn/issues/5428) related to packaging of packages in a monorepo. So I had to think of another solution and this is what I came up with 😁.

**Comment by evocateur** [comment]
One solution that works quite well (with npm) is relative [`file:` specifiers](https://github.com/npm/npm/blob/latest/doc/spec/file-specifiers.md) between leaf packages and `npm i packages/*` in the root. lerna `>=3.0.0-alpha.0` (`npm i -D lerna@next`) supports this seamlessly, resolving relative paths to the correct version range during publish. It works _so_ well, this very repository itself does it.

Symlink approaches like the one you describe have been considered before, many times. They almost always fail because it is too fiddly, has too many edge-cases, and has unintended side-effects with package managers.

**Comment by netanelgilad** [comment]
Thanks @evocateur I tried your solution. I probably didn't fully get it. Am I not supposed to use `lerna bootstrap` any more? Because once I do `npm i packages/*` it creates symlinks to the packages in the root `node_modules` and the packages themselves don't get their own `node_modules` so it somewhat solves the problem but introduces another one. If I have `package-a` depending on `lodash@3.0.0` and `package-b` depending `lodash@4.0.0` the version of lodash that is created in the root `node_modules` is just one of those, and the packages do not get their own correct copy of lodash.
 
I know this issue may have a bigger scope than just lerna, but I'm starting the discussion to find the path forward across lerna and package managers to a solution where developing inside a monorepo doesn't create discrepancies between the way the modules would work in production and in development.

I don't want to take too much of your time, but if you have any references to prior art regarding the kind of solution I suggested, I would love to dive deeper into it. I can try mapping out all the edge cases to see if a solution can be found to handle all known cases.

Let me know If what I said about your suggested solution is wrong in someway. I can open a repository to showcase my issues if that would be clearer than my description.

**Comment by evocateur** [comment]
Yeah, relative file specifiers completely replace lerna bootstrap with npm install in the root. Conflicting majors of dependencies should automatically nest under the leaf directory as a “bundled” dependency; otherwise dependencies are hoisted naturally.

I personally consider sibling packages in a monorepo that have conflicting majors of a dependency to be a bug. They are related, after all, else why bother putting them in the monorepo in the first place?

**Comment by evocateur** [comment]
There’s no shame in developing npm packages “the old-fashioned way” in their own repositories. npm link isn’t terrible, nor is it always required (if the packages are _truly_ independent, that is).

You’re gonna have a bad time when your “truly independent” modules want to version at the same time; despite choosing (potentially) different semver bumps, they are nevertheless coupled to the notion of “I am ready for a release at this revision”. A multi-package repository should not be a grab-bag for random packages. Lerna’s goals do not include “reduce the number of git clones on my hard drive”.
